pandas/io/pytables.py

```diff
@@ -4405,6 +4406,13 @@ def _unconvert_string_array(data, nan_rep=None, encoding=None):
                 dtype = "U{0}".format(itemsize)
             else:
                 dtype = "S{0}".format(itemsize)
+            # fix? issue #10366
+            # it seems like for categoricals, when data.dtype is object the data.astype line does
+            # not raise exception (in python3) and I think this is expected in
+            # order to trigger the decoding in the except below. TODO review.
+            if encoding is not None:
```

Comment:
use ``_convert_string_array(data, _ensure_encoding(encoding))`` (just to reuse things)